homeassistant/components/fritz/sensor.py

```diff
@@ -0,0 +1,151 @@
+        _retrieve_external_ip_state,
+    ],
+    "fritz_uptime": ["Uptime", DEVICE_CLASS_TIMESTAMP, None, _retrieve_uptime_state],
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up entry."""
+    _LOGGER.debug("Setting up FRITZ!Box sensors")
+    fritzbox_tools = hass.data[DOMAIN][entry.entry_id]
+
+    if "WANIPConn1" in fritzbox_tools.connection.services:
+        # Only routers are supported at the moment
+
+        for sensor_type in SENSOR_TYPES_FRITZ:
+            async_add_entities(
+                [
+                    FritzBoxSensor(
+                        fritzbox_tools, entry.title, SENSOR_TYPES_FRITZ[sensor_type]
+                    )
+                ],
+                True,
+            )
+
+
+class FritzBoxSensor(FritzBoxBaseEntity, BinarySensorEntity):
+    """Define FRITZ!Box connectivity class."""
+
+    def __init__(
+        self, fritzbox_tools: FritzBoxTools, device_friendlyname: str, sensor_type: str
+    ) -> None:
+        """Init FRITZ!Box connectivity class."""
+        self._sensor_type = sensor_type
+        self._unique_id = f"{fritzbox_tools.unique_id}-{self._sensor_type[SENSOR_NAME].replace(' ', '_').lower()}"
+        self._name = f"{device_friendlyname} {self._sensor_type[SENSOR_NAME]}"
+        self._is_available = True
+        self._last_value: str | None = None
+        self._state: str | None = None
+        super().__init__(fritzbox_tools, device_friendlyname)
+
+    @property
+    def _state_provider(self):
+        """Return the state provider for the binary sensor."""
+        return self._sensor_type[SENSOR_STATE_PROVIDER]
+
+    @property
+    def name(self):
+        """Return name."""
+        return self._name
+
+    @property
+    def device_class(self) -> str | None:
+        """Return device class."""
+        return self._sensor_type[SENSOR_DEVICE_CLASS]
+
+    @property
+    def icon(self):
+        """Return icon."""
+        return self._sensor_type[SENSOR_ICON]
+
+    @property
+    def unique_id(self):
+        """Return unique id."""
+        return self._unique_id
+
+    @property
+    def state(self) -> str | None:
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def available(self) -> bool:
+        """Return availability."""
+        return self._is_available
+
+    def update(self) -> None:
+        """Update data."""
+        _LOGGER.debug("Updating FRITZ!Box sensors")
+
+        try:
+            status = self._fritzbox_tools.fritzstatus
+            self._is_available = True
+
+            self._state = self._last_value = self._state_provider(
+                status, self._last_value
+            )
+
+        except FritzConnectionException:
+            _LOGGER.error("Error getting the state from the FRITZ!Box", exc_info=True)
+            self._is_available = False
+
+
+def get_device_uptime(uptime: int, last_uptime: str | None) -> str:
```

Comment:
Is this used anywhere else besides `_retrieve_uptime_state`? Can you move the contents of this function into `_retrieve_uptime_state` instead?